File: CS50XR2/pset6/credit/credit.py
# Name: credit.py
# Author: Robin Goyal
# Last-Modified: March 12, 2018
# Purpose: Determine whether a credit card number is valid


import cs50


def main():

    # Get card number
    card_number = cs50.get_int("Number: ")

    # Create array of digits from card number
    card_digits = [int(digit) for digit in reversed(str(card_number))]

    # Calculate sum as specified in Luhn's algorithm
    luhn_sum = 0

    for i in range(len(card_digits)):

        # Sum the individual digits in 2 * digit for every other digit
        if i % 2 == 1:
            luhn_sum += sum(int(digit) for digit in str(2 * card_digits[i]))
        else:
            luhn_sum += card_digits[i]

    print(check_luhn_sum(reversed(card_digits), luhn_sum))


def check_luhn_sum(arr, luhn):
    """
    (list, int) -> str

    Return a string indicating the type of the card
    after checking if the luhn sum is valid and arr
    checking the specific card values
    """

    if luhn % 10 != 0:
        return "INVALID"

    # Cards beginning with 34 or 37 are AMEX
    if arr[0] == 3 and (arr[1] in (4, 7)):
        card_type = "AMEX"

    # Cards beginning with 4 are VISA
    elif arr[0] == 4:
        card_type = "VISA"

    # Remaining cards are mastercard
    else:
        card_type = "MASTERCARD"

    return card_type


if __name__ == "__main__":
    main()

File: cs50/intro-to-cs-round-1/pset1/credit.c
/* Robin Goyal */
/* Determine whether a credit card is valid using Luhn's algorithm */

#include <stdio.h>
#include <cs50.h>
#include <math.h>
#include <string.h>

int main(void) {
    
    // Store user input of credit card number
    
    long long cardNum;
    
    // Obtain a positive value user input
    do {
        printf("Number: ");
        cardNum = get_long_long();
    } 
    while(cardNum < 0);
    
    int nDigits = floor(log10(labs(cardNum))) + 1;
    int cardDigits[nDigits];
    
    // Store cardNum digits in array
    for (int i = 0; i < nDigits; i++) {
        cardDigits[i] = cardNum % 10;
        cardNum = cardNum / 10;
    }
    
    int tmpSum = 0;
    int tmpVal;
    int tmpLength;
    
    // checksum algorithm
    for (int i = 1; i <= nDigits-1; i = i + 2) {
        
        tmpVal = cardDigits[i] * 2;
        tmpLength = floor(log10(abs(tmpVal))) + 1;
        
        for (int j = 0; j < tmpLength; j++){
            tmpSum += tmpVal % 10;
            tmpVal = tmpVal / 10;
        }
    }
    
    for (int i = 0; i < nDigits; i = i + 2) {
        tmpSum += cardDigits[i];
    }
    
    
    // Determines the validity of the card and card type
    if (tmpSum % 10 == 0) {
        
        if ((cardDigits[nDigits-1]) == 4) {
            printf("VISA\n");
        }
        
        else if( (cardDigits[nDigits-1] == 5) && (cardDigits[nDigits-2] == 1 || cardDigits[nDigits-2] == 2 || cardDigits[nDigits-2] == 3 || 
        cardDigits[nDigits-2] == 4 || cardDigits[nDigits-2] == 5) ) {
            printf("MASTERCARD\n"); 
        }
        
        else if (cardDigits[nDigits-1] == 3 && (cardDigits[nDigits-2] == 4 || cardDigits[nDigits-2] == 7)) {
            printf("AMEX\n");
        }
    }
    
    else {
        printf("INVALID\n");
    }
    
    
    return 0;
}